include/gpcc/string/BinaryDumper.hpp

```diff
@@ -0,0 +1,205 @@
+ * dumped. The @ref BinaryDumper instance cannot be reused. After all data has been dumped, the object is intended to be
+ * destroyed.
+ *
+ * The following example will print a dump of binary data to a command line interface:
+ * ~~~{.cpp}
+ * // CLI instantiated somewhere
+ * gpcc::cli::CLI cli;
+ *
+ * // data to be dumped
+ * std::vector<uint8_t> binaryData;
+ * uintptr_t const vaddr = 0x250;
+ *
+ * BinaryDumper bd(vaddr, binaryData.data(), binaryData.Size(), 1U);
+ *
+ * cli.WriteLine(bd.GetHeadLine());
+ *
+ * // Note: If there would be zero bytes of data, then the do-while-loop will ensure that at least one line
+ * // containing the address but no data will be printed.
+ * do
+ * {
+ *   cli.WriteLine(bd.GetLine());
+ * }
+ * while (!bd.IsAllDataDumped());
+ * ~~~
+ *
+ * # Dumping large data in small chunks
+ * The @ref BinaryDumper allows the user to dump large amounts of data (e.g. 1kB) using a small buffer (e.g. 32 byte).
+ *
+ * After all data has been dumped (@ref IsAllDataDumped() returns `true`), the user may provide more data via
+ * @ref ProvideMoreData() and then continue retrieving lines via @ref GetLine() until all of the additional data has
+ * been dumped, too. Multiple small fragments of a large block of binary data can be dumped this way.
+ * @ref ProvideMoreData() will accept new fragments until the last line of the dump was not completely filled with
+ * 16 bytes of data, which is a safe indicator that the whole large block of data has been dumped.
+ *
+ * Example:
+ * ~~~{.cpp}
+ * // CLI instantiated somewhere
+ * gpcc::cli::CLI cli;
+ *
+ * // storage for one fragment of data that shall be dumped
+ * uint8_t data[64];
+ * size_t nBytes;
+ *
+ * // fetch up to 64 byte of data, e.g. from a large file
+ * FetchNext64Byte(data, &nBytes);
+ *
+ * // The address may start at any value. Here we choose zero.
+ * uintptr_t const vaddr = 0x00;
+ *
+ * BinaryDumper bd(vaddr, data, nBytes, 1U);
+ *
+ * // write the headline
+ * cli.WriteLine(bd.GetHeadLine());
+ *
+ * // Dump the data and add more data whenever a fragment is fully dumped.
+ * // If there is no data at all, then the inner do-while-loop will ensure that at least one line containing the address
+ * // but no data will be printed.
+ * do
+ * {
+ *   do
+ *   {
+ *     cli.WriteLine(bd.GetLine());
+ *   }
+ *   while (!bd.IsAllDataDumped());
+ *
+ *   if (bd.IsMoreDataAccepted())
+ *   {
+ *     FetchNext64Byte(data, &nBytes);
+ *     bd.ProvideMoreData(data, nBytes);
+ *   }
+ * }
+ * while (!bd.IsAllDataDumped());
+ * ~~~
+ * - - -
+ *
+ * __Thread safety:__\n
+ * Not thread safe, but non-modifying concurrent access is safe.
+ */
+class BinaryDumper final
+{
+  public:
+    BinaryDumper(void) = delete;
+    BinaryDumper(uintptr_t const address,
+                 uint8_t const wordSize);
+    BinaryDumper(uintptr_t const address,
+                 void const * const pData,
+                 size_t const nBytes,
+                 uint8_t const wordSize);
```

Comment:
The 'input' (binary data) related parameters (pData, nBytes) should be located before the 'output' (string) related parameters (address, wordSize).
```suggestion
    BinaryDumper(void const * const pData,
                 size_t const nBytes,
                 uintptr_t const address,
                 uint8_t const wordSize);
```

Reply:
Yes makes sense, will be adopted.